ouroboros-consensus/src/Ouroboros/Consensus/Util/Enclose.hs

```diff
@@ -0,0 +1,32 @@
+{-# LANGUAGE PatternSynonyms #-}
+
+-- | Utility functions for enclosing a code segment with tracing events.
+module Ouroboros.Consensus.Util.Enclose (
+    Enclosing
+  , Enclosing' (..)
+  , encloseWith
+  , pattern FallingEdge
+  ) where
+
+import           Control.Tracer (Tracer, traceWith)
+
+data Enclosing' a =
+    -- | Preceding a specific code segment.
+    RisingEdge
+    -- | Suceeding a specific code segment, with extra information.
+  | FallingEdge' !a
```

Comment:
Maybe `FallingEdgeWith` instead of the prime? It's not easy, but I try to avoid prime/tick/under etc where possible.

Reply:
Thanks, agreed that a `With` suffix is more indicative than a prime here :+1: